AX: Allow strict replacements via AXTextOperation

Sameesunkaria opened the pull request:
#### c688b31fdd986a6fdbe6643d1d5983042b601af7
<pre>
AX: Allow strict replacements via AXTextOperation
<a href="https://bugs.webkit.org/show_bug.cgi?id=278458">https://bugs.webkit.org/show_bug.cgi?id=278458</a>
<a href="https://rdar.apple.com/134409037">rdar://134409037</a>

Reviewed by Tyler Wilcock.

The AXTextOperation accessibility API for macOS allows text replacement within a specified
list of text marker ranges, among other operations. This AX API is particularly beneficial
for clients like Grammarly for macOS, that utilize the accessibility subsystem to perform
text replacements. However, in its current form, AXTextOperation arbitrarily changes the
case of the replacement string and performs a “smart replacement” that adds spaces around
the replaced text, severely limiting the usability of the API. This commit introduces the
following modifications to the API to allow better control over the replacement. While the
same outcome can be achieved by chaining multiple calls to AX APIs/emitting CGEvents,
AXTextOperation allows for a more efficient implementation.

Preserving the Case of the Replacement String:

`TextOperationReplace` modifies the case of the replacement string supplied to the API,
changing it to either capitalized or lowercase, presumably to match the case of the text
being replaced. As a result, clients of AXTextOperation cannot enforce the case of the
replacement string they provide.

A real example of the existing behavior that makes incorrect/unexpected changes shows up
when the case of the replacement string must be preserved, e.g. when replacing “Mac OS X”
with “macOS”. The current implementation would infer that “Mac OS X” is capitalized and
would attempt to capitalize the replacement string “macOS” -&gt; “MacOS” to preserve the
case.

To address this, an additional `AXTextOperationType` called
&quot;TextOperationReplacePreserveCase&quot; has been introduced, which allows for text replacements
that preserve the case of the replacement string for clients that know/want the
replacement string to maintain its case. When supplying the new type, the replacement
string will be used as is, without attempting to match its case to the string being
replaced. Since this is an additive change, the existing clients of AXTextOperation will
remain unaffected.

Allow Disabling Smart Replacements:

All AXTextOperation types that replace text (i.e. all operations types except “select”),
always perform a “smart replace”, which assumes that the replacement string is not
replacing partial words and attempts to add spaces around it. While this is an excellent
feature for certain replacements, it can severely limit the usability of AXTextOperation
for partial word replacements.

For example, when replacing “MacOS” with “macOS”, we could attempt to perform a minimal
change, where the “M” is replaced with “m” (given we already have
`TextOperationReplacePreserveCase` from above). The current implementation would perform
a smart replacement and, therefore, add a space after the replaced “m”,
“MacOS” -&gt; “m acOS”.

To address this, a new key called &quot;AXTextOperationSmartReplace&quot; has been added to the
AXTextOperation dictionary to support _non-smart_ replacements. The value of this key is
a boolean, where NO/false disables smart replacements. Since this is an additive change,
the existing clients of AXTextOperation will remain unaffected.

Usage:

Client pseudo-code below:
```
// Replace “Mac OS X” -&gt; “macOS”
CFTypeRef result;
AXUIElementCopyParameterizedAttributeValue(element, &quot;AXTextOperation&quot;, @[
  NSAccessibilityTextOperationType: NSAccessibilityTextOperationReplacePreserveCase,
  NSAccessibilityTextOperationReplacementString: @&quot;macOS&quot;,
  NSAccessibilityTextOperationMarkerRanges: @[ rangeOfMacOSX ]
], &amp;result)

// Replace &quot;MacOS&quot; -&gt; &quot;macOS&quot;, by only replacing the first character
CFTypeRef result;
AXUIElementCopyParameterizedAttributeValue(element, &quot;AXTextOperation&quot;, @[
  NSAccessibilityTextOperationType: NSAccessibilityTextOperationReplacePreserveCase,
  NSAccessibilityTextOperationReplacementString: @&quot;m&quot;,
  NSAccessibilityTextOperationSmartReplace: kCFBooleanFalse,
  NSAccessibilityTextOperationMarkerRanges: @[ rangeOfM ]
], &amp;result)
```

Layout tests have been introduced for the various AXTextOperation types. At the time of
writing, AXTextOperation has a bug with replacements/transformations over multiple
ranges, which is why some of the tests cases are expected to fail until the bug is
resolved. See <a href="https://bugs.webkit.org/show_bug.cgi?id=278928">https://bugs.webkit.org/show_bug.cgi?id=278928</a>

* LayoutTests/accessibility/mac/text-operation/text-operation-capitalize-expected.txt: Added.
* LayoutTests/accessibility/mac/text-operation/text-operation-capitalize.html: Added.
* LayoutTests/accessibility/mac/text-operation/text-operation-lowercase-expected.txt: Added.
* LayoutTests/accessibility/mac/text-operation/text-operation-lowercase.html: Added.
* LayoutTests/accessibility/mac/text-operation/text-operation-replace-expected.txt: Added.
* LayoutTests/accessibility/mac/text-operation/text-operation-replace-preserve-case-expected.txt: Added.
* LayoutTests/accessibility/mac/text-operation/text-operation-replace-preserve-case.html: Added.
* LayoutTests/accessibility/mac/text-operation/text-operation-replace.html: Added.
* LayoutTests/accessibility/mac/text-operation/text-operation-select-expected.txt: Added.
* LayoutTests/accessibility/mac/text-operation/text-operation-select.html: Added.
* LayoutTests/accessibility/mac/text-operation/text-operation-smart-replace-expected.txt: Added.
* LayoutTests/accessibility/mac/text-operation/text-operation-smart-replace.html: Added.
* LayoutTests/accessibility/mac/text-operation/text-operation-uppercase-expected.txt: Added.
* LayoutTests/accessibility/mac/text-operation/text-operation-uppercase.html: Added.
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WebCore/accessibility/AXCoreObject.h:
(WebCore::AccessibilityTextOperation::AccessibilityTextOperation): Deleted.
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::performTextOperation):
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(accessibilityTextOperationForParameterizedAttribute):
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityTextMarkerRange.cpp:
(WTR::toTextMarkerRange):
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityTextMarkerRange.h:
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityUIElement.cpp:
(WTR::AccessibilityUIElement::performTextOperation):
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityUIElement.h:
* Tools/WebKitTestRunner/InjectedBundle/Bindings/AccessibilityUIElement.idl:
* Tools/WebKitTestRunner/InjectedBundle/mac/AccessibilityUIElementMac.mm:
(WTR::textOperationParameterizedAttribute):
(WTR::AccessibilityUIElement::performTextOperation):

Canonical link: <a href="https://commits.webkit.org/283740@main">https://commits.webkit.org/283740@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/573c135a49aa7994322c0a58d03de02be4a2783e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/66909 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/46284 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/19531 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/70944 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/18042 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/69027 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/54083 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/17823 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/53664 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/12143 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/69976 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/42581 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/57891 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/34208 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/39252 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/15283 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/16396 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/61170 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/15624 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/72645 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/10866 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/14974 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/61075 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/10898 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/57948 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/61231 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14902 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/8926 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/2542 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/42091 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/43168 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/44351 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/42911 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebCore/accessibility/AXCoreObject.h:
```diff
@@ -640,16 +640,21 @@ enum class AccessibilityTextOperationType {
 
     AccessibilityTextOperation()
         : type(AccessibilityTextOperationType::Select)
+        , smartReplace(AccessibilityTextOperationSmartReplace::Yes)
```
Rather than setting the default in the constructor, let's set it explicitly in line with the member definition:

`AccessibilityTextOperationSmartReplace smartReplace { AccessibilityTextOperationSmartReplace::Yes };`

Sameesunkaria replied:
Also moved the default for the operation `type` to maintain consistency.

twilco commented on Source/WebCore/accessibility/AccessibilityObject.cpp:
```diff
@@ -1047,14 +1047,17 @@ Vector<String> AccessibilityObject::performTextOperation(const AccessibilityText
         // return the replacement string when replacing, but the selected string if not.
         if (replaceSelection) {
-            frame->editor().replaceSelectionWithText(replacementString, Editor::SelectReplacement::Yes, Editor::SmartReplace::Yes);
+            frame->editor().replaceSelectionWithText(replacementString, Editor::SelectReplacement::Yes, static_cast<Editor::SmartReplace>(operation.smartReplace));
```
I think I'd prefer to be more explicit and avoid `static_cast`, even if it's a bit more verbose. How about:

```cpp
frame->editor().replaceSelectionWithText(replacementString, Editor::SelectReplacement::Yes, operation.smartReplace == AccessibilityTextOperationSmartReplace::No ? Editor::SmartReplace::No : Editor::SmartReplace::Yes);
```

Sameesunkaria replied:
That’s where I started, but the verbosity took over! 😅 I'll make the switch—clarity wins!

Sameesunkaria replied:
Updated

twilco reviewed (commented):
Looks good to me overall! Ideally we would want a layout test to go along with this change to ensure we don't break this in the future. But unfortunately, there are no existing tests of this functionality, nor the `AccessibilityUIElement` infrastructure to even write such a test 😞 

Here's what we'd need to do:

1. In `AccessibilityUIElement.idl`, add something like:

```cpp
object searchTextWithCriteria(object searchStrings, DOMString startFrom, DOMString direction);
// new line
object performTextOperation(object markerRanges, DOMString operationType, DOMString replacement, boolean shouldSmartReplace);
```

2. Add this method to the `WebKitTestRunner/InjectedBundle/AccessibilityUIElement.h`, and implement it in `WebKitTestRunner/InjectedBundle/Mac/AccessibilityUIElementMac.mm`. We can use `#if PLATFORM(MAC)` guards like `searchTextWithCriteria`, since this is a Mac-only API. `AccessibilityUIElement::searchTextWithCriteria` is a good model for how we would implement this new function. Don't worry about changing anything in the `DumpRenderTree` folder (that is for WebKit 1), just `WebKitTestRunner`.

3. Write a new layout test, e.g. `LayoutTests/accessibility/mac/smart-replace-text-operation.html`. I imagine a testcase where we have two contenteditables or something with identical strings, and for one we do a smart replacement, and the other we do a non-smart replacement, and verify both results are what we expect. [hidden-slot.html](https://github.com/WebKit/WebKit/blob/main/LayoutTests/accessibility/hidden-slot.html) is a good example of a simple layout test with modern style we'd expect, vs. something like [mac/search-text/search-text.html](https://github.com/WebKit/WebKit/blob/main/LayoutTests/accessibility/mac/search-text/search-text.html) which is more similar in spirit to the text operations API, but has an outdated test style (e.g. `shouldBe` instead of `expect`).

4. You can run the test using `Tools/Scripts/run-webkit-tests --no-build --no-retry accessibility/mac/smart-replace-text-operation.html`.

Please let me know here or in the WebKit Slack if you have questions.

Sameesunkaria commented on Tools/WebKitTestRunner/InjectedBundle/AccessibilityTextMarkerRange.cpp:
```diff
@@ -70,5 +70,10 @@ JSClassRef AccessibilityTextMarkerRange::wrapperClass()
+
+AccessibilityTextMarkerRange* toTextMarkerRange(JSObjectRef object)
+{
+    return static_cast<AccessibilityTextMarkerRange*>(JSObjectGetPrivate(object));
```
Perhaps there is a better way of doing this, like by directly using `toAccessibilityTextMarkerRange` from `JSAccessibilityTextMarkerRange.h` at the call site, but I don't see any of the generated JS wrapper classes being included in files other than in the implementation of the wrapped classes.

twilco replied:
I looked around and couldn't find a better way. All other usages of `JSObjectGetPrivate` are paired with a `static_cast`, and the same function you added here also exists in `DumpRenderTree`, so there's precedence. Let's go with this.

twilco commented on LayoutTests/accessibility/mac/text-operation/text-operation-uppercase.html:
```diff
@@ -0,0 +1,81 @@
+<p contenteditable="true" id="text3">TEXT3: The quick brown fox jumps over the lazy dog.</p>
+
+<script>
+    description("Tests that an Uppercase text operation transforms text to uppercase as expected.");
```
Two small nits for this test and others:

  1. Typically, rather than using `description`, we'd do:

```js
var output = "Tests that an Uppercase text operation transforms text to uppercase as expected.\n\n";
```

And then you don't need the other `var output` declaration below.

  2. Everything in the `<script>` tag can move back one indentation level — helps avoid a bit of rightward drift.

twilco commented on LayoutTests/accessibility/mac/text-operation/text-operation-smart-replace.html:
```diff
@@ -0,0 +1,92 @@
+<script>
+    description("Tests that a text operation replacement produces the expected result with and without smart-replacement.");
+
+    function markerRangeForSpanById(id, element) {
```
One important thing to note: WebKit accessibility has two "modes" — one that serves all accessibility requests from the main-thread, and another that serves accessibility requests from a dedicated accessibility thread with a cached view of the page.

This matters here because it takes time for the accessibility thread cache to update when the DOM changes, like via our usage of `range.selectNodeContents(document.getElementById(id));`. And if we don't wait for the cache to update, this test is going to fail in "accessibility thread" mode because the `selectedTextMarkerRange` won't have updated by the time we ask for it. This is not a problem in the "normal" main-thread mode because every query goes directly to the DOM.

You can run tests in this accessibility thread mode by passing `--accessibility-isolated-tree` to the `run-webkit-tests` script (there was no way for you to know this 🙂 — tests in this configuration run post-commit only at the moment). All of these great new tests fail in this mode, but we can fix that!

1. All of these tests will need to set `window.jsTestIsAsync = true;` — this is typically done in the first line inside `if (window.accessibilityController)`.
2. All tests will need `finishJSTest();` after `debug(output);` (otherwise the async-test infrastructure never knows when to stop). I think you have this on all tests already, so all good.
3. Inside `if (window.accessibilityController)`, before the first use of `await`, wrap all remaining test content like so:

```js
setTimeout(async function() {
    // We're in an async block, so we're allowed to await.

    debug(output);
    finishJSTest();
}, 0);
```

A fair amount of other tests do this, so feel free to look around for examples.

4. I propose changing `markerRangeForSpanById(id, element)` to be something like this:

```js
// Select the text of the element for the given id. Returns the selected text marker range
// after the accessibility tree has updated with the new selection.
async function selectElementTextById(id) {
    window.getSelection().clear();
    // Wait until the selection is cleared so we start from a known state.
    await waitFor(() => !accessibilityController.rootElement.selectedTextMarkerRange());
    ...insert all the code you wrote select the node contents...
    let selectedRange;
    await waitFor(() => {
        selectedRange = accessibilityController.rootElement.selectedTextMarkerRange();
        // Wait until we have a selected text marker range. Because we started by clearing the selection,
        // we know that any non-null selected range must be for the right element.
        return selectedRange;
    });
    return selectedRange;
}
```

Then in the test, we can use it similarly to how it is currently, e.g.:

```js
..., [await selectElementTextById("target1"), await selectElementTextById("target2")], ...
```

Finally, this function seems like a good candidate to add to [accessibility-helper.js](https://github.com/WebKit/WebKit/blob/main/LayoutTests/resources/accessibility-helper.js), so we don't need to duplicate it every test.

Sameesunkaria replied:
I was actually wondering why we needed the `waitFor` in several existing tests, cause to me it seemed like calls were only returning after completing the respective operation. Thank you for the explanation.

twilco commented on LayoutTests/accessibility/mac/text-operation/text-operation-smart-replace.html:
```diff
@@ -0,0 +1,92 @@
+        return element.selectedTextMarkerRange();
+    }
+
+    function markerRangeInElement(id, start, end) {
```
Maybe `startIndex` and `endIndex` vs. `start` and `end` to make it obvious these are number values? This function will need to be made async, too.

twilco reviewed (commented):
First, thank you very much for adding so much test coverage, and for the side investigation in https://bugs.webkit.org/show_bug.cgi?id=278928! Looks very good overall, just a few more things to work out.

I built your patch locally to try it out, and I'm getting strange whitespace failures on these tests:

```
Regressions: Unexpected text-only failures (4)
accessibility/mac/text-operation/text-operation-capitalize.html [ Failure ]
accessibility/mac/text-operation/text-operation-lowercase.html [ Failure ]
accessibility/mac/text-operation/text-operation-smart-replace.html [ Failure ]
accessibility/mac/text-operation/text-operation-uppercase.html [ Failure ]
```

I attached a screenshot of the raw diff, and the pretty diff of `text-operation-uppercase.html` (it's the same story for all four of these tests). The test `PASS`es, but the whitespace is different than expected. It seems to have something to do with how smart replace inserts non-breaking-spaces.

<img width="1256" alt="Screenshot of text-operation-uppercase-diff.html. The key point made by the image is that in the expectation that reads 'jumps ov...', my run of the test generates an 'Â' unicode character, which is not expected. But the test is still a PASS." src="https://github.com/user-attachments/assets/cad9c960-6966-46b6-929d-5ff3261dfa3f">

<img width="1338" alt="Screenshot of text-operation-uppercase-pretty-diff.html. The key point of the image is that the coloring of the diff indicates differences between the file in the whitespace separating 'ov' and 'er', even though it's just a whitespace in both cases." src="https://github.com/user-attachments/assets/abfc6f39-38ff-42fe-becc-52b7bb2e68a3">

I've also attached the diff files in case you'd like to examine the actual contents.

[text-operation-uppercase-diff.txt](https://github.com/user-attachments/files/16857263/text-operation-uppercase-diff.txt)

[text-operation-uppercase-pretty-diff.txt](https://github.com/user-attachments/files/16857284/text-operation-uppercase-pretty-diff.txt) (Change this file extension to .html — GitHub doesn't allow the uploading of .html files)

I see you included `\xa0` in some expectations, so I'm guessing you ran into weirdness here too.

Not sure why this would differ between my system, and yours + the bots (since the tests pass). I'm running macOS Sonoma 14.6.1, while the bots run Ventura.

In any case, we can avoid this problem by checking only for expected substrings, instead of full strings.

```js
output += expect("text.stringValue", "'AXValue: The quick BROWN FOx jumps ov\xa0ER THE LA zy dog.'");
vs.
output += expect("text.stringValue.includes('value that proves we behave correctly')", "true");
```

What do you think about this? Will we still be able to test expected behavior by checking for substrings only?

If so, we'll need to avoid the whitespace differences in the rendered text after `TEST COMPLETE`, too. The text after `TEST COMPLETE` is all the text on the page when the test completes, so we can hide it to avoid this noise.

```html
<div id="test-content">
    <p contenteditable="true" id="text">...</p>
    <p contenteditable="true" id="text2">...</p>
    <p contenteditable="true" id="text3">...</p>
</div>
<script>
...test logic...

document.getElementById("test-content").visibility = "hidden";
debug(output);
finishJSTest();
</script>
```

---

Open to discuss any other ideas on solving this if you have some other insight here! Thanks again for your hard work.

Sameesunkaria commented on LayoutTests/accessibility/mac/text-operation/text-operation-capitalize.html:
```diff
@@ -0,0 +1,111 @@
+        output += expect("operationResult[2]", "'E Quick Bro'");
+        output += expect("text.stringValue.replace(/\\s/g, ' ')", "'AXValue: TEXT3: Th E Quick Bro wn fox jum Ps Over the Lazy dog.'");
+
+        document.getElementById("test-content").remove();
```
It seems like a hidden element still appears in the expected text, so I removed it altogether.

Sameesunkaria commented on LayoutTests/accessibility/mac/text-operation/text-operation-capitalize.html:
```diff
@@ -0,0 +1,111 @@
+
+    accessibilityElement.addNotificationListener(listener);
+    await operation();
+    await waitFor(() => { return reached; });
```
I ran into flakiness when using a Promise directly, where the element's stringValue hadn't updated. A waitFor seems more reliable (I ran it 1000 times without failure).

Please let me know if I should take a different approach, like waiting on the expected value. Wanted to implement something along the lines of this https://github.com/WebKit/WebKit/pull/33440#discussion_r1754776076

twilco commented on LayoutTests/accessibility/mac/text-operation/text-operation-capitalize.html:
```diff
@@ -0,0 +1,111 @@
+<script>
+var output = "Tests that a Capitalize text operation transforms text to capitalized as expected.\n\n";
+
+async function getMarkerRangeInElementById(id, startIndex, endIndex) {
```
Maybe we could add this to `accessibility-helpers.js`? The name should reflect that it's performing a selection as a side-effect, too. Maybe `selectPartialElementTextById(id, startIndex, endIndex, axWebArea)`, where the last parameter allows the caller to optionally pass in the webarea `AccessibilityUIElement` so we can retrieve it only once, rather than every time this function is called (see [docs for selectElementTextById](https://github.com/WebKit/WebKit/blob/64e0879489042142d935d6b615999b111bc1a5e1/LayoutTests/resources/accessibility-helper.js#L270#L272)).

twilco commented on LayoutTests/accessibility/mac/text-operation/text-operation-capitalize.html:
```diff
@@ -0,0 +1,111 @@
+    return selectedRange;
+}
+
+async function waitForNotification(accessibilityElement, notification, operation) {
```
Nice, this will be really useful for future tests! Maybe we could add it to `accessibility-helpers.js` so we don't need copies of it in all of the new tests?

Would be great to add docs when we do that too, especially for the `operation` parameter. Seems useful to spell out that it's excepted to be a function (does it have to be `async`? If so let's note that) that should produce the expected notification when run.

Maybe `notificationName` rather than `notification`?

Sameesunkaria replied:
I wanted to avoid polluting `accessibility-helpers.js`, but I do agree that it would be nice to have this function for future tests. I've moved it. 🙂

> does it have to be async?

It doesn't _need_ to be async; it was slightly more ergonomic for my specific use case, but now I'm calling it as a normal function instead because it felt more appropriate.

twilco reviewed (approved):
Looking great, almost there!

Sameesunkaria committed:
AX: Allow strict replacements via AXTextOperation
https://bugs.webkit.org/show_bug.cgi?id=278458
rdar://134409037

Reviewed by Tyler Wilcock.

The AXTextOperation accessibility API for macOS allows text replacement within a specified
list of text marker ranges, among other operations. This AX API is particularly beneficial
for clients like Grammarly for macOS, that utilize the accessibility subsystem to perform
text replacements. However, in its current form, AXTextOperation arbitrarily changes the
case of the replacement string and performs a “smart replacement” that adds spaces around
the replaced text, severely limiting the usability of the API. This commit introduces the
following modifications to the API to allow better control over the replacement. While the
same outcome can be achieved by chaining multiple calls to AX APIs/emitting CGEvents,
AXTextOperation allows for a more efficient implementation.

Preserving the Case of the Replacement String:

`TextOperationReplace` modifies the case of the replacement string supplied to the API,
changing it to either capitalized or lowercase, presumably to match the case of the text
being replaced. As a result, clients of AXTextOperation cannot enforce the case of the
replacement string they provide.

A real example of the existing behavior that makes incorrect/unexpected changes shows up
when the case of the replacement string must be preserved, e.g. when replacing “Mac OS X”
with “macOS”. The current implementation would infer that “Mac OS X” is capitalized and
would attempt to capitalize the replacement string “macOS” -> “MacOS” to preserve the
case.

To address this, an additional `AXTextOperationType` called
"TextOperationReplacePreserveCase" has been introduced, which allows for text replacements
that preserve the case of the replacement string for clients that know/want the
replacement string to maintain its case. When supplying the new type, the replacement
string will be used as is, without attempting to match its case to the string being
replaced. Since this is an additive change, the existing clients of AXTextOperation will
remain unaffected.

Allow Disabling Smart Replacements:

All AXTextOperation types that replace text (i.e. all operations types except “select”),
always perform a “smart replace”, which assumes that the replacement string is not
replacing partial words and attempts to add spaces around it. While this is an excellent
feature for certain replacements, it can severely limit the usability of AXTextOperation
for partial word replacements.

For example, when replacing “MacOS” with “macOS”, we could attempt to perform a minimal
change, where the “M” is replaced with “m” (given we already have
`TextOperationReplacePreserveCase` from above). The current implementation would perform
a smart replacement and, therefore, add a space after the replaced “m”,
“MacOS” -> “m acOS”.

To address this, a new key called "AXTextOperationSmartReplace" has been added to the
AXTextOperation dictionary to support _non-smart_ replacements. The value of this key is
a boolean, where NO/false disables smart replacements. Since this is an additive change,
the existing clients of AXTextOperation will remain unaffected.

Usage:

Client pseudo-code below:
```
// Replace “Mac OS X” -> “macOS”
CFTypeRef result;
AXUIElementCopyParameterizedAttributeValue(element, "AXTextOperation", @[
  NSAccessibilityTextOperationType: NSAccessibilityTextOperationReplacePreserveCase,
  NSAccessibilityTextOperationReplacementString: @"macOS",
  NSAccessibilityTextOperationMarkerRanges: @[ rangeOfMacOSX ]
], &result)

// Replace "MacOS" -> "macOS", by only replacing the first character
CFTypeRef result;
AXUIElementCopyParameterizedAttributeValue(element, "AXTextOperation", @[
  NSAccessibilityTextOperationType: NSAccessibilityTextOperationReplacePreserveCase,
  NSAccessibilityTextOperationReplacementString: @"m",
  NSAccessibilityTextOperationSmartReplace: kCFBooleanFalse,
  NSAccessibilityTextOperationMarkerRanges: @[ rangeOfM ]
], &result)
```

Layout tests have been introduced for the various AXTextOperation types. At the time of
writing, AXTextOperation has a bug with replacements/transformations over multiple
ranges, which is why some of the tests cases are expected to fail until the bug is
resolved. See https://bugs.webkit.org/show_bug.cgi?id=278928

* LayoutTests/accessibility/mac/text-operation/text-operation-capitalize-expected.txt: Added.
* LayoutTests/accessibility/mac/text-operation/text-operation-capitalize.html: Added.
* LayoutTests/accessibility/mac/text-operation/text-operation-lowercase-expected.txt: Added.
* LayoutTests/accessibility/mac/text-operation/text-operation-lowercase.html: Added.
* LayoutTests/accessibility/mac/text-operation/text-operation-replace-expected.txt: Added.
* LayoutTests/accessibility/mac/text-operation/text-operation-replace-preserve-case-expected.txt: Added.
* LayoutTests/accessibility/mac/text-operation/text-operation-replace-preserve-case.html: Added.
* LayoutTests/accessibility/mac/text-operation/text-operation-replace.html: Added.
* LayoutTests/accessibility/mac/text-operation/text-operation-select-expected.txt: Added.
* LayoutTests/accessibility/mac/text-operation/text-operation-select.html: Added.
* LayoutTests/accessibility/mac/text-operation/text-operation-smart-replace-expected.txt: Added.
* LayoutTests/accessibility/mac/text-operation/text-operation-smart-replace.html: Added.
* LayoutTests/accessibility/mac/text-operation/text-operation-uppercase-expected.txt: Added.
* LayoutTests/accessibility/mac/text-operation/text-operation-uppercase.html: Added.
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WebCore/accessibility/AXCoreObject.h:
(WebCore::AccessibilityTextOperation::AccessibilityTextOperation): Deleted.
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::performTextOperation):
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(accessibilityTextOperationForParameterizedAttribute):
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityTextMarkerRange.cpp:
(WTR::toTextMarkerRange):
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityTextMarkerRange.h:
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityUIElement.cpp:
(WTR::AccessibilityUIElement::performTextOperation):
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityUIElement.h:
* Tools/WebKitTestRunner/InjectedBundle/Bindings/AccessibilityUIElement.idl:
* Tools/WebKitTestRunner/InjectedBundle/mac/AccessibilityUIElementMac.mm:
(WTR::textOperationParameterizedAttribute):
(WTR::AccessibilityUIElement::performTextOperation):

Canonical link: https://commits.webkit.org/283740@main
